File: packages/contracts/src/payment.ts
import { z } from "zod";

import { dataEnvelopeSchema } from "./envelope";

export const paymentStatusSchema = z.enum(["PENDING", "VERIFIED", "REJECTED"]);

export type PaymentStatus = z.infer<typeof paymentStatusSchema>;

export const createPaymentInputSchema = z
  .object({
    invoiceId: z.string().uuid(),
    amount: z.number().int().positive(),
    paidAt: z.string().regex(/^\d{4}-\d{2}-\d{2}$/, "Tanggal harus menggunakan format YYYY-MM-DD."),
    note: z.string().trim().min(3).max(300),
  })
  .strict();

export type CreatePaymentInput = z.infer<typeof createPaymentInputSchema>;

export const rejectPaymentInputSchema = z
  .object({
    reason: z.string().trim().min(3).max(500),
  })
  .strict();

export type RejectPaymentInput = z.infer<typeof rejectPaymentInputSchema>;

export const paymentListQuerySchema = z.object({
  status: paymentStatusSchema.optional(),
  limit: z.coerce.number().int().min(1).max(100).default(20),
});

export const paymentSchema = z.object({
  id: z.string().uuid(),
  invoiceId: z.string().uuid(),
  duesTypeName: z.string().min(1),
  period: z.string().min(1),
  amount: z.number().int().nonnegative(),
  paidAt: z.string(),
  note: z.string().min(1),
  status: paymentStatusSchema,
  submittedByName: z.string().min(1),
  houseCode: z.string().min(1),
  householdDisplayName: z.string().min(1),
  verifiedByName: z.string().nullable(),
  verifiedAt: z.string().datetime({ offset: true }).nullable(),
  rejectionReason: z.string().nullable(),
  receiptNumber: z.string().nullable(),
  createdAt: z.string().datetime({ offset: true }),
});

export type Payment = z.infer<typeof paymentSchema>;

export const paymentListResponseSchema = dataEnvelopeSchema(
  z.object({ items: z.array(paymentSchema) }),
);

export const paymentMutationResponseSchema = dataEnvelopeSchema(
  z.object({ payment: paymentSchema }),
);

export type PaymentListResponse = z.infer<typeof paymentListResponseSchema>;
export type PaymentMutationResponse = z.infer<typeof paymentMutationResponseSchema>;

File: apps/mobile/lib/features/invoice/domain/payment.dart
enum PaymentStatus {
  pending('PENDING'),
  verified('VERIFIED'),
  rejected('REJECTED');

  const PaymentStatus(this.apiValue);

  final String apiValue;

  static PaymentStatus fromApi(Object? value) {
    if (value is! String) {
      throw const FormatException('Payment status is invalid.');
    }
    for (final status in values) {
      if (status.apiValue == value) return status;
    }
    throw FormatException('Unsupported payment status: $value');
  }
}

String paymentStatusLabel(PaymentStatus status) {
  switch (status) {
    case PaymentStatus.pending:
      return 'Menunggu Verifikasi';
    case PaymentStatus.verified:
      return 'Terverifikasi';
    case PaymentStatus.rejected:
      return 'Ditolak';
  }
}

/// Mirrors `Payment` in `packages/contracts/src/payment.ts`.
class Payment {
  const Payment({
    required this.id,
    required this.invoiceId,
    required this.duesTypeName,
    required this.period,
    required this.amount,
    required this.paidAt,
    required this.note,
    required this.status,
    required this.submittedByName,
    required this.houseCode,
    required this.householdDisplayName,
    required this.verifiedByName,
    required this.verifiedAt,
    required this.rejectionReason,
    required this.receiptNumber,
    required this.createdAt,
  });

  final String id;
  final String invoiceId;
  final String duesTypeName;
  final String period;
  final int amount;

  /// YYYY-MM-DD, matching the API's date-only contract.
  final String paidAt;
  final String note;
  final PaymentStatus status;
  final String submittedByName;
  final String houseCode;
  final String householdDisplayName;
  final String? verifiedByName;
  final DateTime? verifiedAt;
  final String? rejectionReason;
  final String? receiptNumber;
  final DateTime createdAt;

  factory Payment.fromJson(Map<String, dynamic> json) {
    final verifiedAtValue = json['verifiedAt'];
    return Payment(
      id: json['id'] as String,
      invoiceId: json['invoiceId'] as String,
      duesTypeName: json['duesTypeName'] as String,
      period: json['period'] as String,
      amount: (json['amount'] as num).toInt(),
      paidAt: json['paidAt'] as String,
      note: json['note'] as String,
      status: PaymentStatus.fromApi(json['status']),
      submittedByName: json['submittedByName'] as String,
      houseCode: json['houseCode'] as String,
      householdDisplayName: json['householdDisplayName'] as String,
      verifiedByName: json['verifiedByName'] as String?,
      verifiedAt: verifiedAtValue == null
          ? null
          : DateTime.parse(verifiedAtValue as String),
      rejectionReason: json['rejectionReason'] as String?,
      receiptNumber: json['receiptNumber'] as String?,
      createdAt: DateTime.parse(json['createdAt'] as String),
    );
  }
}
